x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyMaintenanceService.java

```diff
@@ -0,0 +1,173 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.common.xcontent.ObjectPath;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.enrich.EnrichPolicy;
+
+public class EnrichPolicyMaintenanceService implements LocalNodeMasterListener {
+
+    private static final Logger logger = LogManager.getLogger(EnrichPolicyMaintenanceService.class);
+
+    private static final String MAPPING_POLICY_FIELD_PATH = MapperService.SINGLE_MAPPING_NAME + "._meta." +
+        EnrichPolicyRunner.ENRICH_POLICY_FIELD_NAME;
+    private static final IndicesOptions IGNORE_UNAVAILABLE = IndicesOptions.fromOptions(true, false, false, false);
+
+    private final Settings settings;
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final EnrichPolicyLocks enrichPolicyLocks;
+
+    private volatile Scheduler.Cancellable cancellable;
+
+    EnrichPolicyMaintenanceService(Settings settings, Client client, ClusterService clusterService, ThreadPool threadPool,
+                                   EnrichPolicyLocks enrichPolicyLocks) {
+        this.settings = settings;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.enrichPolicyLocks = enrichPolicyLocks;
+    }
+
+    void initialize() {
+        clusterService.addLocalNodeMasterListener(this);
+    }
+
+    @Override
+    public void onMaster() {
+        if (cancellable == null || cancellable.isCancelled()) {
+            scheduleNext();
+            clusterService.addLifecycleListener(new LifecycleListener() {
+                @Override
+                public void beforeStop() {
+                    offMaster();
+                }
+            });
+        }
+    }
+
+    @Override
+    public void offMaster() {
+        if (cancellable != null && cancellable.isCancelled() == false) {
+            cancellable.cancel();
+        }
+    }
+
+    @Override
+    public String executorName() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    private void scheduleNext() {
+        try {
+            TimeValue waitTime = EnrichPlugin.ENRICH_CLEANUP_PERIOD.get(settings);
+            cancellable = threadPool.schedule(this::execute, waitTime, ThreadPool.Names.GENERIC);
+        } catch (EsRejectedExecutionException e) {
+            if (e.isExecutorShutdown()) {
+                logger.debug("failed to schedule next [enrich] maintenance task; shutting down", e);
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    private void execute() {
+        logger.debug("triggering scheduled [enrich] maintenance task");
```

Comment:
The cleanup period defaults to 15 minutes, so the likelihood that two cleanups run at the time is small.

But maybe we can check if `cancellable` is set and if it is then not execute `cleanUpEnrichIndices ()` method?

We would also need to unset `cancellable` after old indices have been removed or it was determined that no indices need to be deleted.